### PR TITLE
build: rewrite release script

### PR DESCRIPTION
**Pull request by keithamus**
@manuelpuyol this updates #292, rewriting the release script a little, such that you can pick to bump via Major, Minor, Version. It also ensures npm is in lockstep with the ruby gem, which we want to happen, I think.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "@primer/view-components",
-  "version": "0.0.1",
+  "name": "@primer/view_components",
+  "version": "0.0.20",
```
Comment (manuelpuyol) [inline package.json:3]:
```suggestion
  "version": "0.0.21",
```

File: script/release [hunk 2]
```diff
@@ -1,43 +1,105 @@
...
+    demo/Gemfile.lock \
+    lib/primer/view_components/version.rb \
+    app/assets/javascripts \
+    package.json \
+    yarn.lock
+}
+
+commit() {
+  git commit -m "release $1.$2.$3"
+  git tag "v$1.$2.$3"
```
Comment (manuelpuyol) [inline script/release:51]:
```suggestion
```

Tagging will be handled after merging when publishing the gem

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ Here are a few things you can do that will increase the likelihood of your pull
 
 If you are the current maintainer of this gem:
 
+1. Checkout to `main` and `git pull`
```
Comment (srt32) [inline CONTRIBUTING.md:34]: We could also make the script do this part? Make sure it's on `main` and then `git pull` and confirm no conflicts?

File: script/release [hunk 4]
```diff
@@ -1,43 +1,105 @@
...
+}
+
+push() {
```
Comment (manuelpuyol) [inline script/release:54]:
```suggestion
push() {
  git push origin release

```

File: script/publish [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# Publish NPM
+read -p "Publish NPM (y/N)? " answer
+case ${answer:0:1} in
+    y|Y )
+        npm publish --access public
+    ;;
+    * )
+        # noop
+    ;;
+esac
```
Comment (manuelpuyol) [inline script/publish:23]:
```suggestion
bundle exec rake release

# Publish NPM
npm publish --access public
```

**Review by manuelpuyol — COMMENTED** [review]
The only "issue" I see with this approach is that we'll create a bunch of empty JS releases, since most of our gem releases don't change any JS

File: script/release [hunk 6]
```diff
@@ -1,43 +1,105 @@
...
+  echo "####################################################"
+  echo "Now, open a PR with this branch and merge it to main"
+  echo "Then, run script/publish on main to release the gem and the NPM package"
+  echo "Finally, create a GitHub release https://github.com/primer/view_components/releases/new with the changes from CHANGELOG"
```
Comment (srt32) [inline script/release:57]: I wonder if we could automate even more of this if we were okay relying the `gh` CLI?

**Comment by manuelpuyol** [comment]
@srt32 I wonder if we could create an Action similar to the docs build that runs this (adapted) script for us

File: script/release [hunk 7]
```diff
@@ -1,43 +1,105 @@
...
+    echo new version $major $minor $patch
+
+    create_release_branch
+    update_readme $major $minor $patch
+    update_ruby_version $major $minor $patch
+    update_gemfiles $major $minor $patch
+    update_npm $major $minor $patch
+    add_changed_files
+    commit
+    push
```
Comment (srt32) [inline script/release:102]: Do we not need `--tags` here?

Comment (keithamus) [reply script/release:102]: Just removed the tagging behaviour, so not any more.

File: script/release (hunk 6 above)
Comment (manuelpuyol) [reply script/release:59]: Yes! That's a next step in my opinion
we could use gh API to create the release and everything

File: script/release (hunk 7 above)
Comment (manuelpuyol) [reply script/release:102]: https://github.com/primer/view_components/pull/296#discussion_r584891031

**Comment by srt32** [comment]
> @srt32 I wonder if we could create an Action similar to the docs build that runs this (adapted) script for us

We certainly could! We'd need to be comfy putting creds in there though, right?

**Comment by manuelpuyol** [comment]
> > @srt32 I wonder if we could create an Action similar to the docs build that runs this (adapted) script for us
> 
> We certainly could! We'd need to be comfy putting creds in there though, right?

oh yeah, forgot about that
also we need 2FA to publish the gem 😅 
I'll look more into this later, I want to be as close as possible to how https://github.com/primer/doctocat releases their changes

File: lib/primer/view_components/version.rb [hunk 8]
```diff
@@ -11,3 +11,7 @@ module VERSION
     end
   end
 end
+
+if __FILE__ == $0
+  puts Primer::ViewComponents::VERSION::STRING
+end
```
Comment (manuelpuyol) [inline lib/primer/view_components/version.rb:17]:
```suggestion
puts Primer::ViewComponents::VERSION::STRING if __FILE__ == $PROGRAM_NAME
```

let's make rubocop happy 😄

**Comment by srt32** [comment]
> > > @srt32 I wonder if we could create an Action similar to the docs build that runs this (adapted) script for us
> > 
> > 
> > We certainly could! We'd need to be comfy putting creds in there though, right?
> 
> oh yeah, forgot about that
> also we need 2FA to publish the gem 😅
> I'll look more into this later, I want to be as close as possible to how https://github.com/primer/doctocat releases their changes

Is https://github.com/primer/doctocat/blob/master/.github/workflows/release.yml#L29 not using 2FA? I think we want it to be...

**Comment by manuelpuyol** [comment]
> > > > @srt32 I wonder if we could create an Action similar to the docs build that runs this (adapted) script for us
> > > 
> > > 
> > > We certainly could! We'd need to be comfy putting creds in there though, right?
> > 
> > 
> > oh yeah, forgot about that
> > also we need 2FA to publish the gem 😅
> > I'll look more into this later, I want to be as close as possible to how https://github.com/primer/doctocat releases their changes
> 
> Is https://github.com/primer/doctocat/blob/master/.github/workflows/release.yml#L29 not using 2FA? I think we want it to be...

I'm not sure how they release, but doesn't look like they are using 2FA there

**Review by manuelpuyol — APPROVED** [review]
amazing @keithamus !

File: script/release [hunk 9]
```diff
@@ -1,43 +1,165 @@
...
+  fetch
+  if ! remote_history_is_clean; then 
+    echo "Error: changes exist on origin not pulled into this branch. Please pull"
+    exit 1
+  fi 
+
+  if ! local_history_is_clean; then 
+    echo "Error: changes exist that haven't been pushed to origin. Please pull"
+    exit 1
+  fi
```
Comment (keithamus) [inline script/release:120]: So it now checks:

 - You're on `main`
 - You have a clean tree, no erroneous files
 - You have pushed up everything
 - You've pulled down everything
 - You're not going to overwrite an existing tag

I've tried to test these as best I can but I guess we'll try running this for real when merged and see how it goes ;)

Comment (manuelpuyol) [reply script/release:120]: Having some errors in this script is fine, we can start using it and fix it as it goes

**Pull request merged**